File: lessons/09-pattern-matching/pattern_matching.js
#!/usr/bin/env node

/**
 * Lesson 9: Pattern Matching in JavaScript
 *
 * JavaScript doesn't have native pattern matching (yet - there's a proposal),
 * but we can simulate it with various techniques:
 * - Object destructuring
 * - Switch statements with fall-through
 * - Guard patterns with if/else
 * - Libraries like ts-pattern
 *
 * This demonstrates pattern-like techniques available in JavaScript.
 */

// ====================
// 1. Basic Switch (Limited Pattern Matching)
// ====================

function describeNumber(n) {
    switch (n) {
        case 0: return "zero";
        case 1: return "one";
        case 2: return "two";
        default: return `many: ${n}`;
    }
}

// ====================
// 2. Destructuring (Tuple-like Patterns)
// ====================

function describePoint([x, y]) {
    if (x === 0 && y === 0) return "origin";
    if (x === 0) return `on y-axis at y=${y}`;
    if (y === 0) return `on x-axis at x=${x}`;
    return `point at (${x}, ${y})`;
}

function describeList(lst) {
    const [first, second, ...rest] = lst;

    if (lst.length === 0) return "empty list";
    if (lst.length === 1) return `single element: ${first}`;
    if (lst.length === 2) return `two elements: ${first}, ${second}`;
    return `first: ${first}, rest: [${rest.join(', ')}]`;
}

// ====================
// 3. Object Destructuring
// ====================

function processCommand(cmd) {
    const { action, direction, target, damage } = cmd;

    switch (action) {
        case "quit":
            return "Quitting...";
        case "move":
            return `Moving ${direction}`;
        case "attack":
            return `Attacking ${target} for ${damage} damage`;
        default:
            return `Unknown action: ${action}`;
    }
}

// ====================
// 4. Type-Based Pattern Simulation
// ====================

class Shape {
    getType() { return "Shape"; }
}

class Circle extends Shape {
    constructor(radius) {
        super();
        this.radius = radius;
    }
    getType() { return "Circle"; }
}

class Rectangle extends Shape {
    constructor(width, height) {
        super();
        this.width = width;
        this.height = height;
    }
    getType() { return "Rectangle"; }
}

class Triangle extends Shape {
    constructor(a, b, c) {
        super();
        this.a = a;
        this.b = b;
        this.c = c;
    }
    getType() { return "Triangle"; }
}

function area(shape) {
    if (shape instanceof Circle) {
        return Math.PI * shape.radius * shape.radius;
    } else if (shape instanceof Rectangle) {
        return shape.width * shape.height;
    } else if (shape instanceof Triangle) {
        const { a, b, c } = shape;
        const s = (a + b + c) / 2;
        return Math.sqrt(s * (s - a) * (s - b) * (s - c));
    }
    throw new Error("Unknown shape");
}

// ====================
// 5. Guards (Conditional Patterns)
// ====================

function classify(n) {
    if (n < 0) return "negative";
    if (n === 0) return "zero";
    if (n < 10) return "small positive";
    if (n < 100) return "medium positive";
    return "large positive";
}

// ====================
// 6. Multiple Values (OR patterns)
// ====================

function isWeekend(day) {
    const d = day.toLowerCase();
    return d === "saturday" || d === "sunday";
}

// ====================
// 7. Pattern Matching Helper Function
// ====================

function match(value, patterns) {
    for (const [predicate, result] of patterns) {
        if (typeof predicate === 'function') {
            if (predicate(value)) {
                return typeof result === 'function' ? result(value) : result;
            }
        } else if (predicate === value) {
            return typeof result === 'function' ? result(value) : result;
        }
    }
    throw new Error(`No pattern matched for value: ${value}`);
}

function gradeLetterUsingMatch(score) {
    return match(score, [
        [s => s >= 90 && s <= 100, 'A'],
        [s => s >= 80 && s < 90, 'B'],
        [s => s >= 70 && s < 80, 'C'],
        [s => s >= 60 && s < 70, 'D'],
        [s => s < 60, 'F']
    ]);
}

// ====================
// 8. Optional/Nullable Pattern
// ====================

function describeMaybe(opt) {
    if (opt === null || opt === undefined) {
        return "nothing";
    }
    return `just ${opt}`;
}

function describeResult(result) {
    if (result.ok) {
        return `success: ${result.value}`;
    }
    return `error: ${result.error}`;
}

// ====================
// 9. Expression Evaluator
// ====================

class Expr {
    eval() { throw new Error("Abstract method"); }
}

class Num extends Expr {
    constructor(value) {
        super();
        this.value = value;
    }
    eval() { return this.value; }
}

class Add extends Expr {
    constructor(left, right) {
        super();
        this.left = left;
        this.right = right;
    }
    eval() { return this.left.eval() + this.right.eval(); }
}

class Mul extends Expr {
    constructor(left, right) {
        super();
        this.left = left;
        this.right = right;
    }
    eval() { return this.left.eval() * this.right.eval(); }
}

class Neg extends Expr {
    constructor(expr) {
        super();
        this.expr = expr;
    }
    eval() { return -this.expr.eval(); }
}

// ====================
// 10. Nested Patterns
// ====================

function analyzeNested(data) {
    const { user, active, error } = data;

    if (error) {
        return `Error: ${error}`;
    }

    if (user) {
        const { name, age } = user;
        if (active) {
            return `Active user: ${name}${age ? ` (${age})` : ''}`;
        }
        return `Inactive user: ${name}`;
    }

    return "Unknown data";
}

// ====================
// 11. State Machine
// ====================

function trafficLightNext(current, action) {
    if (action === "emergency") return "red";

    const transitions = {
        "red": { "timer": "green" },
        "green": { "timer": "yellow" },
        "yellow": { "timer": "red" }
    };

    return transitions[current]?.[action] || current;
}

// ====================
// 12. Advanced Pattern Helper (Visitor-like)
// ====================

function matchObject(obj, patterns) {
    for (const pattern of patterns) {
        if (pattern.when(obj)) {
            return pattern.then(obj);
        }
    }
    return patterns.find(p => p.default)?.then(obj) || null;
}

function processUser(userData) {
    return matchObject(userData, [
        {
            when: u => u.type === "admin",
            then: u => `Admin ${u.name} with ${u.permissions?.length || 0} permissions`
        },
        {
            when: u => u.type === "user" && u.age >= 18,
            then: u => `Adult user: ${u.name}`
        },
        {
            when: u => u.type === "user",
            then: u => `Minor user: ${u.name}`
        },
        {
            when: u => u.type === "guest",
            then: u => `Guest session: ${u.session}`
        },
        {
            default: true,
            when: () => true,
            then: () => "Unknown user type"
        }
    ]);
}

// ====================
// Main Demonstration
// ====================

function main() {
    console.log("=== Pattern Matching in JavaScript ===\n");

    // 1. Basic switch
    console.log("1. Basic Switch Patterns:");
    [0, 1, 2, 5].forEach(n => {
        console.log(`   describeNumber(${n}) = ${describeNumber(n)}`);
    });

    // 2. Destructuring patterns
    console.log("\n2. Destructuring (Tuple-like) Patterns:");
    const points = [[0, 0], [0, 5], [3, 0], [2, 3]];
    points.forEach(p => {
        console.log(`   [${p}] -> ${describePoint(p)}`);
    });

    // 3. List patterns
    console.log("\n3. List Patterns:");
    const lists = [[], [1], [1, 2], [1, 2, 3, 4]];
    lists.forEach(lst => {
        console.log(`   [${lst}] -> ${describeList(lst)}`);
    });

    // 4. Object patterns
    console.log("\n4. Object Patterns:");
    const commands = [
        { action: "quit" },
        { action: "move", direction: "north" },
        { action: "attack", target: "orc", damage: 10 }
    ];
    commands.forEach(cmd => {
        console.log(`   ${JSON.stringify(cmd)}`);
        console.log(`   -> ${processCommand(cmd)}`);
    });

    // 5. Type-based patterns
    console.log("\n5. Type-Based Patterns (instanceof):");
    const shapes = [
        new Circle(5),
        new Rectangle(4, 6),
        new Triangle(3, 4, 5)
    ];
    shapes.forEach(shape => {
        console.log(`   ${shape.getType()}: area = ${area(shape).toFixed(2)}`);
    });

    // 6. Guards
    console.log("\n6. Guards (Conditional Patterns):");
    [-5, 0, 3, 50, 500].forEach(n => {
        console.log(`   ${n} -> ${classify(n)}`);
    });

    // 7. OR patterns
    console.log("\n7. Multiple Values (OR patterns):");
    ["Monday", "Saturday", "Sunday", "Wednesday"].forEach(day => {
        console.log(`   ${day} is weekend? ${isWeekend(day)}`);
    });

    // 8. Match helper
    console.log("\n8. Custom Match Helper:");
    [95, 85, 75, 65, 55].forEach(score => {
        console.log(`   ${score} -> ${gradeLetterUsingMatch(score)}`);
    });

    // 9. Optional patterns
    console.log("\n9. Optional/Nullable Patterns:");
    console.log(`   ${describeMaybe(42)}`);
    console.log(`   ${describeMaybe(null)}`);
    console.log(`   ${describeResult({ ok: true, value: 100 })}`);
    console.log(`   ${describeResult({ ok: false, error: "failed" })}`);

    // 10. Expression evaluator
    console.log("\n10. Expression Evaluator:");
    const expr = new Mul(new Add(new Num(2), new Num(3)), new Num(4));
    console.log(`   (2 + 3) * 4 = ${expr.eval()}`);
    const expr2 = new Neg(new Add(new Num(5), new Num(3)));
    console.log(`   -(5 + 3) = ${expr2.eval()}`);

    // 11. Nested patterns
    console.log("\n11. Nested Patterns:");
    const nestedData = [
        { user: { name: "Alice", age: 30 }, active: true },
        { user: { name: "Bob" }, active: false },
        { error: "Not found" }
    ];
    nestedData.forEach(data => {
        console.log(`   ${analyzeNested(data)}`);
    });

    // 12. State machine
    console.log("\n12. State Machine (Traffic Light):");
    let state = "red";
    ["timer", "timer", "timer", "emergency"].forEach(action => {
        state = trafficLightNext(state, action);
        console.log(`   After '${action}': ${state}`);
    });

    // 13. Advanced pattern matching
    console.log("\n13. Advanced Object Matching:");
    const users = [
        { type: "admin", name: "Alice", permissions: ["read", "write"] },
        { type: "user", name: "Bob", age: 25 },
        { type: "user", name: "Charlie", age: 16 },
        { type: "guest", session: "abc123" }
    ];
    users.forEach(user => {
        console.log(`   ${processUser(user)}`);
    });

    console.log("\n=== JavaScript Pattern Matching Notes ===");
    console.log("- No native pattern matching (yet - TC39 proposal in progress)");
    console.log("- Use destructuring for simple patterns");
    console.log("- instanceof for type-based patterns");
    console.log("- Custom helper functions for complex matching");
    console.log("- Libraries like ts-pattern provide full pattern matching");
}

main();

File: lessons/09-pattern-matching/pattern_matching.rs
// Lesson 9: Pattern Matching in Rust
//
// Rust has excellent pattern matching! It's exhaustive by default,
// meaning the compiler ensures you handle all possible cases.

// ====================
// 1. Basic Match Expression
// ====================

fn describe_number(n: i32) -> &'static str {
    match n {
        0 => "zero",
        1 => "one",
        2 => "two",
        _ => "many",  // _ is wildcard
    }
}

// ====================
// 2. Tuple Patterns
// ====================

fn describe_point(point: (i32, i32)) -> String {
    match point {
        (0, 0) => "origin".to_string(),
        (0, y) => format!("on y-axis at y={}", y),
        (x, 0) => format!("on x-axis at x={}", x),
        (x, y) => format!("point at ({}, {})", x, y),
    }
}

// ====================
// 3. Enum Patterns
// ====================

enum Shape {
    Circle { radius: f64 },
    Rectangle { width: f64, height: f64 },
    Triangle { a: f64, b: f64, c: f64 },
}

fn area(shape: &Shape) -> f64 {
    use std::f64::consts::PI;
    match shape {
        Shape::Circle { radius } => PI * radius * radius,
        Shape::Rectangle { width, height } => width * height,
        Shape::Triangle { a, b, c } => {
            let s = (a + b + c) / 2.0;
            (s * (s - a) * (s - b) * (s - c)).sqrt()
        }
    }
}

// ====================
// 4. Option Pattern Matching
// ====================

fn describe_option(opt: Option<i32>) -> String {
    match opt {
        Some(x) => format!("just {}", x),
        None => "nothing".to_string(),
    }
}

// ====================
// 5. Result Pattern Matching
// ====================

fn describe_result(res: Result<i32, String>) -> String {
    match res {
        Ok(val) => format!("success: {}", val),
        Err(e) => format!("error: {}", e),
    }
}

// ====================
// 6. Guards
// ====================

fn classify(n: i32) -> &'static str {
    match n {
        n if n < 0 => "negative",
        0 => "zero",
        n if n < 10 => "small positive",
        n if n < 100 => "medium positive",
        _ => "large positive",
    }
}

// ====================
// 7. Or Patterns
// ====================

fn is_weekend(day: &str) -> bool {
    match day.to_lowercase().as_str() {
        "saturday" | "sunday" => true,
        _ => false,
    }
}

// ====================
// 8. Range Patterns
// ====================

fn grade_letter(score: u32) -> char {
    match score {
        90..=100 => 'A',
        80..=89 => 'B',
        70..=79 => 'C',
        60..=69 => 'D',
        _ => 'F',
    }
}

// ====================
// 9. Expression Evaluator
// ====================

enum Expr {
    Num(i32),
    Add(Box<Expr>, Box<Expr>),
    Mul(Box<Expr>, Box<Expr>),
    Neg(Box<Expr>),
}

fn eval(expr: &Expr) -> i32 {
    match expr {
        Expr::Num(n) => *n,
        Expr::Add(left, right) => eval(left) + eval(right),
        Expr::Mul(left, right) => eval(left) * eval(right),
        Expr::Neg(e) => -eval(e),
    }
}

// ====================
// 10. Nested Patterns
// ====================

fn analyze_nested(data: (Option<i32>, Option<i32>)) -> String {
    match data {
        (Some(x), Some(y)) => format!("both: {} and {}", x, y),
        (Some(x), None) => format!("only first: {}", x),
        (None, Some(y)) => format!("only second: {}", y),
        (None, None) => "neither".to_string(),
    }
}

// ====================
// 11. Struct Patterns
// ====================

#[derive(Debug)]
struct User {
    name: String,
    age: u32,
    active: bool,
}

fn describe_user(user: &User) -> String {
    match user {
        User { active: true, age, name } if *age >= 18 => {
            format!("Active adult: {}", name)
        }
        User { active: true, name, .. } => {
            format!("Active minor: {}", name)
        }
        User { active: false, name, .. } => {
            format!("Inactive: {}", name)
        }
    }
}

// ====================
// Main Demonstration
// ====================

fn main() {
    println!("=== Pattern Matching in Rust ===\n");

    // 1. Basic matching
    println!("1. Basic Match:");
    for n in [0, 1, 2, 5] {
        println!("   {} -> {}", n, describe_number(n));
    }

    // 2. Tuple patterns
    println!("\n2. Tuple Patterns:");
    let points = vec![(0, 0), (0, 5), (3, 0), (2, 3)];
    for p in points {
        println!("   {:?} -> {}", p, describe_point(p));
    }

    // 3. Enum patterns
    println!("\n3. Enum Patterns (Shapes):");
    let shapes = vec![
        Shape::Circle { radius: 5.0 },
        Shape::Rectangle { width: 4.0, height: 6.0 },
        Shape::Triangle { a: 3.0, b: 4.0, c: 5.0 },
    ];
    for shape in &shapes {
        println!("   area = {:.2}", area(shape));
    }

    // 4. Option
    println!("\n4. Option Pattern:");
    println!("   {}", describe_option(Some(42)));
    println!("   {}", describe_option(None));

    // 5. Result
    println!("\n5. Result Pattern:");
    println!("   {}", describe_result(Ok(100)));
    println!("   {}", describe_result(Err("failed".to_string())));

    // 6. Guards
    println!("\n6. Guards:");
    for n in [-5, 0, 3, 50, 500] {
        println!("   {} -> {}", n, classify(n));
    }

    // 7. Or patterns
    println!("\n7. Or Patterns:");
    for day in ["Monday", "Saturday", "Sunday"] {
        println!("   {} is weekend? {}", day, is_weekend(day));
    }

    // 8. Range patterns
    println!("\n8. Range Patterns:");
    for score in [95, 85, 75, 65, 55] {
        println!("   {} -> {}", score, grade_letter(score));
    }

    // 9. Expression evaluator
    println!("\n9. Expression Evaluator:");
    let expr = Expr::Mul(
        Box::new(Expr::Add(Box::new(Expr::Num(2)), Box::new(Expr::Num(3)))),
        Box::new(Expr::Num(4)),
    );
    println!("   (2 + 3) * 4 = {}", eval(&expr));

    // 10. Nested patterns
    println!("\n10. Nested Patterns:");
    let cases = vec![
        (Some(1), Some(2)),
        (Some(3), None),
        (None, Some(4)),
        (None, None),
    ];
    for case in cases {
        println!("   {:?} -> {}", case, analyze_nested(case));
    }

    // 11. Struct patterns
    println!("\n11. Struct Patterns:");
    let users = vec![
        User { name: "Alice".to_string(), age: 25, active: true },
        User { name: "Bob".to_string(), age: 16, active: true },
        User { name: "Charlie".to_string(), age: 30, active: false },
    ];
    for user in &users {
        println!("   {}", describe_user(user));
    }
}
